packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -0,0 +1,288 @@
+
+  let selectedKeyItem = selectedKey ? selectState.collection.getItem(selectedKey) : undefined;
+  let selectedKeyText = selectedKeyItem ? selectedKeyItem.textValue || selectedKeyItem.rendered as string : undefined;
+  // Maybe don't need to do for defaultSelectedKey? Need to do for selectedKey so that the textfield is properly controlled reflects selectedKey text
+  let defaultSelectedKeyItem = defaultSelectedKey ? selectState.collection.getItem(defaultSelectedKey) : undefined;
+  let defaultSelectedKeyText = defaultSelectedKeyItem ? defaultSelectedKeyItem.textValue || defaultSelectedKeyItem.rendered as string : undefined;
+  // Double check if selectedKey should make textfield value controlled
+  let [value, setValue] = useControlledState(toString(inputValue) || selectedKeyText, toString(defaultInputValue) || defaultSelectedKeyText || '', onInputChange);
+  let lowercaseValue = value.toLowerCase();
+
+  let defaultFilterFn = useMemo(() => (node: Node<T>) => {
+    let scan = 0;
+    let lowercaseNode = node.textValue.toLowerCase();
+    for (let i in lowercaseValue) {
+      if (whitespace.test(lowercaseValue[i])) {
+        continue;
+      }
+      let match = false;
+      for (; scan < lowercaseNode.length && !match; scan++) {
+        if (!whitespace.test(lowercaseNode[scan])) { // if whitespace, move on to next character
+          let compareVal = collator.compare(lowercaseValue[i], lowercaseNode[scan]);
+          if (compareVal !== 0) { // if it didn't match, we can exit
+            return false;
+          } else { // we matched, move to next character in input for comparison
+            match = true;
+          }
+        }
+      }
+    }
+    return true;
+  }, [collator, lowercaseValue]);
+
+  useEffect(() => {
+    if (onFilter) {
+      onFilter(value);
+    }
+  // Having onFilter in the dep array seems to break it
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [value]);
+
+  selectState.collection = useMemo(() => {
+    if (itemsControlled || value === '') {
+      return selectState.collection;
+    }
+    return new FilteredCollection(selectState.collection, defaultFilterFn);
+  }, [selectState.collection, value, itemsControlled, defaultFilterFn]);
+
+  useEffect(() => {
+    if (selectState.isOpen && selectState.collection.size === 0) {
+      selectState.close();
+    }
+  }, [selectState.isOpen, selectState.collection, selectState]);
+
+  // Maybe we can get rid of these force parameters if we have the selectState.open/toggle calls
+  // within a useEffect instead (it calls those only if isFocused and isOpen). These func are changed so it calls
+  // setOpen(true) and setOpen(!isOpen)?
+  // Feels kinda convoluted now, prob gona refactor to the above
+  let open = (force = false) => {
+    // Only open if field is focused or force and there are items to display
+    if ((isFocused || force) && selectState.collection.size > 0) {
+      selectState.open();
+    }
+  };
```

Comment:
Feels complicated/convoluted after a couple iterations and there are cases where there will be a `close` and `open` call one after the other (like when you first type in a letter and it doesn't match any thing in the list. I'm thinking of doing the commented refactor and seeing if it makes it more clean (only concern is if the open -> close might still happen depending if the collection filter doesn't happen fast enough and thus we get a erroneous open)